File: src/test/scala/demo/customer/LondonCustomerTests.scala
package demo.customer

import org.scalamock.scalatest.MockFactory
import org.scalatest.flatspec.AnyFlatSpec

class LondonCustomerTests extends AnyFlatSpec with MockFactory {
  it should "purchase successfully when enough inventory" in {
    // Arrange
    val productType = ProductType.Book
    val quantity = 6
    val storeStub = stub[Store]

    (storeStub.hasEnoughInventory _)
      .when(productType, quantity)
      .returns(true)

    // Act
    val updatedStore =
      CustomerService.purchase(storeStub, productType, quantity)

    // Assert
    (storeStub.removeInventory _).verify(productType, quantity).once()
  }

  it should "fail to purchase when not enough inventory" in {
    // Arrange
    val productType = ProductType.Book
    val quantity = 6
    val storeStub = stub[Store]

    (storeStub.hasEnoughInventory _)
      .when(productType, quantity)
      .returns(false)

    // Act
    val updatedStore =
      CustomerService.purchase(storeStub, productType, quantity)

    // Assert
    (storeStub.removeInventory _).verify(productType, quantity).never()
  }
}

File: csharp/CleanTests.Tests/Schools/LondonCustomerTests.cs
using System;
using CleanTests.Customer;
using FluentAssertions;
using Moq;
using Xunit;
using static CleanTests.Customer.ProductType;

namespace CleanTests.Tests.Schools;

public class LondonCustomerTests
{
    private const int Quantity = 6;
    private readonly Mock<IStore> _storeMock;

    public LondonCustomerTests()
    {
        _storeMock = new Mock<IStore>();
    }

    [Fact]
    public void It_Should_Purchase_Successfully_When_Enough_Inventory()
    {
        _storeMock
            .Setup(s => s.HasEnoughInventory(Book, Quantity))
            .Returns(true);

        var updatedStore = CustomerService.Purchase(_storeMock.Object, Book, 6);

        updatedStore.IsSucc().Should().BeTrue();
        _storeMock.Verify(s => s.RemoveInventory(Book, Quantity), Times.Once);
    }

    [Fact]
    public void It_Should_Fail_When_Not_Enough_Inventory()
    {
        _storeMock
            .Setup(s => s.HasEnoughInventory(Book, Quantity))
            .Returns(false);

        var updatedStore = CustomerService.Purchase(_storeMock.Object, Book, 11);

        updatedStore.IsFail().Should().BeTrue();
        updatedStore
            .FailureUnsafe()
            .Should()
            .BeOfType<ArgumentException>()
            .Which
            .Message
            .Should().Be("Not enough inventory");
        _storeMock.Verify(s => s.RemoveInventory(Book, Quantity), Times.Never);
    }
}